pkg/lwip/contrib/sock/ip/lwip_sock_ip.c

```diff
@@ -189,62 +199,146 @@ ssize_t sock_ip_recv_aux(sock_ip_t *sock, void *data, size_t max_len,
     ssize_t res, ret = 0;
     bool nobufs = false;
 
+    bool peek = false;
+    if (aux != NULL) {
+        if (IS_USED(MODULE_SOCK_AUX_PEEK)) {
+            peek = aux->flags & SOCK_AUX_PEEK;
+        }
+    }
+
     assert((sock != NULL) && (data != NULL) && (max_len > 0));
-    while ((res = sock_ip_recv_buf_aux(sock, &pkt, (void **)&ctx, timeout,
-                                       remote, aux)) > 0) {
+    while ((res = sock_ip_recv_buf_aux(sock, &pkt, (void**)&ctx, timeout, remote,
+                                       aux)) > 0) {
+
+        if (peek && IS_USED(MODULE_SOCK_AUX_PEEK)) {
+            // SOCK_AUX_PEEK was unset in previous call of sock_ip_recv_buf_aux
+            aux->flags |= SOCK_AUX_PEEK;
+        }
+
         if (ctx->p->tot_len > (ssize_t)max_len) {
+            /* copy everything that fits in to the buffer and return -ENOBUFS
+             * (quite similar to the MSG_TRUNC flag) */
+            memcpy(ptr, pkt, max_len);
```

Comment:
What if this is the 2nd chunk and `ptr != data`?

Reply:
I'm not sure what you mean with a second chunk. `while (netbuf_next(ctx) == 0) {}` will move to the end. Or do you mean that the while loop could call `sock_ip_recv_buf_aux` a second time? We maybe could conditionally call `break` when `peek` instead of `continue`.

Reply:
In theory `sock_ip_recv_buf_aux()` could return multiple chunks of payload data, that's why it's called in a loop.

But in practice this is never the case - it also makes no sense if the API is supposed to be used for zero-copy operations, since that constraint would require a re-assembly buffer on the client side.

NanoCoAP just [asserts that there can never be a second chunk](https://github.com/RIOT-OS/RIOT/blob/master/sys/net/application_layer/nanocoap/sock.c#L172).